CorsixTH/Src/iso_fs.cpp

```diff
@@ -70,33 +74,231 @@ constexpr int max_directory_depth = 16;
+    iso_directory_iterator()=delete;
+
+    /**
+     * Initialize an iterator for the directory table in the memory region
+     * defined by by begin and end.
+     *
+     * \param begin pointer to the first byte of the directory table.
+     * \param end pointer to the first byte following the directory table.
+     */
+    iso_directory_iterator(const uint8_t* begin, const uint8_t* end) {
+        directory_ptr = begin;
+        end_ptr = end;
+        if (directory_ptr == end_ptr) {
+            // dummy value, not accessible.
+            entry = iso_file_entry();
+        } else {
+            entry = iso_file_entry(begin);
+        }
+    }
+
+    /**
+     * Copy the given iso_directory_iterator
+     */
+    iso_directory_iterator(iso_directory_iterator& it) {
+        directory_ptr = it.directory_ptr;
+        end_ptr = it.end_ptr;
+        entry = it.entry;
+    }
+
+    /**
+     * Move the given iso_directory_iterator
+     */
+    iso_directory_iterator(iso_directory_iterator&& it) {
+        directory_ptr = it.directory_ptr;
+        end_ptr = it.end_ptr;
+        entry = std::move(it.entry);
+        it.directory_ptr = nullptr;
+        it.end_ptr = nullptr;
+        it.entry = iso_file_entry();
+    }
+
+    ~iso_directory_iterator()=default;
+
+    /**
+     * Determine whether two iso_directory_iterators point to the same table
+     * entry.
+     */
+    bool operator==(const iso_directory_iterator& rhs) const {
+        return (this->directory_ptr == rhs.directory_ptr);
+    }
+
+    /**
+     * Determine whether to iso_directory_iterators do not point to the same
+     * table entry.
+     */
+    bool operator!=(const iso_directory_iterator& rhs) const {
+        return !((*this)==rhs);
+    }
+
+    /**
+     * Get the file entry pointed to by the iterator.
+     */
+    reference operator*() const {
+        if (directory_ptr == end_ptr) {
+            throw std::out_of_range("iso directory iterator is past end of input");
+        }
+        return entry;
+    }
+
+    /**
+     * Assign this iterator the value of another iterator by copy
+     */
+    iso_directory_iterator& operator=(iso_directory_iterator& rhs) {
+        directory_ptr = rhs.directory_ptr;
+        end_ptr = rhs.end_ptr;
+        entry = rhs.entry;
+        return *this;
+    }
+
+    /**
+     * Assign this iterator the value of another iterator by move
+     */
+    iso_directory_iterator& operator=(iso_directory_iterator&& rhs) {
+        directory_ptr = rhs.directory_ptr;
+        end_ptr = rhs.end_ptr;
+        entry = std::move(rhs.entry);
+        rhs.directory_ptr = nullptr;
+        rhs.end_ptr = nullptr;
+        rhs.entry = {};
+        return *this;
+    }
+
+    /**
+     * Advance this iterator to the next file entry in the directory table,
+     * returning the result.
+     */
+    iso_directory_iterator& operator++() {
+        directory_ptr += *directory_ptr;
+        while (*directory_ptr == 0 && directory_ptr < end_ptr) {
```

Comment:
Swap the conditions to avoid accessing memory that you're not supposed to access